### ISSUE TITLE
Update unique IDs without breaking translations?

### ISSUE THREAD

**Issue by brylie** [gadicc/meteor-messageformat#105]
Is there a way to update the template helper unique IDs without having to re-translate all the strings? E.g. I started adding template IDs before coming up with a convention, and now need to change some of the original IDs to match the established convention.

**Comment by gadicc** [comment]
Not in the UI, but you could update the `mfStrings` database directly, and also mfExtracts.js and mfAll.js.

**Comment by brylie** [comment]
Cool. I will manually replace them.

Would a fuzzy search be able to suggest when IDs change, such as by checking line number and text value?

**Comment by gadicc** [comment]
Mmm, it's something we could look at but the idea is that IDs should never change :)  We looked into doing fuzzy matching like gettext instead of using IDs, which is possible, but just not always reliable, especially with translations from a database, etc.  In the end we chose the most robust solution but this kind of functionality is still on the wishlist for an optional behaviour.

**Comment by brylie** [comment]
Right, I am just thinking that IDs might need to change in some circumstances. E.g. I named my IDs to include the name of their template (such as *blogPost-submitButtonText*). If I refactor the template into smaller units, I will probably change the ID to match the sub-template name instead.

**Comment by gadicc** [comment]
Yeah, I agree...  would be nice to have an easy way to rename all occurrences of a key (definitely in the database, and maybe even in user's source code).  Will leave this open as an enhancement for v2.

**Comment by brylie** [comment]
I am not sure where else to bring this up, but what are your thoughts on the [Meteor core i18n draft proposal](https://meteor.hackpad.com/i18n-in-Meteor-m9MhtHa8DR2)?

**Comment by gadicc** [comment]
Thanks for the invite to that.  Looks interesting, a lot of reading :)  I added a note about meteor-messageformat, it sounds like he's interested in doing his own implementation though.  I'm going to carry on working on v2 which servers my needs and is based on my own past i18n experiences... when I'm done with my list I'll see how it compares to the proposal.

**Comment by brylie** [comment]
@gadicc it seems like our [initial keys are remaining in the translation file](https://github.com/Crowducate/crowducate-next/pull/145). Are the old keys supposed to be removed when running mf_extract? We have a distributed translation effort, and would like to verify commits, but need more clarification about the file structure.

**Comment by gadicc** [comment]
Hey @brylie.  `mf_extract` will add a `removed: true` property to keys in mfExtract.js that no longer exist in your code.  (As a heads up, v2 doesn't use mf_extract / mfExtract.js anymore).

You're right, this isn't reflected properly in mfAll.js.  We do in fact want to keep old key info in this file for auditing purposes, but it should be tagged appropriately and never sent back to the client.  I'll be working on this on the v2 branch.

In the long term, part of this project's goal is to eliminate updating translations via commit.  Can I confirm that the ideal for you here would be for new translations to be marked as "pending" and only be sent out to the client after an admin has approved that single translation or all pending translations by language / author?

**Comment by gadicc** [comment]
And yes, `mfAll.js` is loaded on startup and can indeed be used to update the local database in the way you describe.

**Comment by brylie** [comment]
We are basically translating an open source application by crowdsourcing the effort. Translators may or may not be familiar with Github, Git, node, etc, so the simpler we can make the process the better. So far, we have been reviewing all work created for the project, including wireframes, code, etc, and the translations would ideally be subject to a review process.
